components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/ApiInvocationHandler.java

```diff
@@ -0,0 +1,169 @@
+        if (fullPath.contains(TOKEN_ENDPOINT) || fullPath.contains(USERINFO_ENDPOINT)) {
+            String basicAuth = (String) headerMap.get(AUTH_HEADER);
+            headerMap.remove(AUTH_HEADER);
+            messageContext.setProperty(TEMP_AUTH_HEADER, basicAuth);
+            clientId = getClientIdForTokenCall(fullPath, basicAuth);
+            if (clientId == null)
+                getClientIdForUserInfoCall(fullPath, basicAuth);
+
+        }
+        return clientId;
+
+    }
+
+    private String getClientIdForUserInfoCall(String fullPath, String basicAuth) {
+        String accessToken = null;
+        if (fullPath.contains(TOKEN_ENDPOINT)) {
+            if (log.isDebugEnabled()) {
+                log.debug(USERINFO_ENDPOINT);
+            }
+            accessToken = APIManagerDBUtil.getAccessTokenForUserInfo(basicAuth.split(" ")[1]);
+        }
+        return accessToken;
+
+    }
+
+    private String getClientIdForTokenCall(String fullPath, String basicAuth) {
+        String clientId = null;
+        if (fullPath.contains(TOKEN_ENDPOINT)) {
+            if (log.isDebugEnabled()) {
+                log.debug(TOKEN_ENDPOINT);
+            }
+            byte[] valueDecoded = Base64.decodeBase64(basicAuth.split(" ")[1].getBytes());
+            String decodeString = new String(valueDecoded);
+            clientId = decodeString.split(":")[0];
+        }
+        return clientId;
+    }
+
+    private boolean processResponse(String clientId, Map headerMap, String fullPath) {
+
+        if (clientId == null)
+            return false;
+
+        else if (fullPath.contains(USERINFO_ENDPOINT)) {
+            headerMap.put(AUTH_HEADER, TOKEN_TYPE + clientId);
+            return true;
+
+        }
+
+        TokenDTO tokenDTO = APIManagerDBUtil.getTokenDetailsFromAPIManagerDB(clientId);
+        headerMap.put(AUTH_HEADER, TOKEN_TYPE + tokenDTO.getAccessToken());
+        handleByTokenPool(clientId, tokenDTO);
+        return true;
+
+    }
+
+    private void handleByTokenPool(String clientId, TokenDTO tokenDTO) {
+        Map<String, String> propertyMap = ReadPropertyFile.getPropertyFile();
+        if (propertyMap.get(TOKEN_POOL_ENABLED).equals("true")) {
+            if (log.isDebugEnabled()) {
+                log.debug("TOken Pool Service Enabled for Insertion");
+            }
+            AddNewSpDTO newSpDto = new AddNewSpDTO();
+            ArrayList<TokenDTO> tokenList = new ArrayList<TokenDTO>();
+
+            newSpDto.setOwnerId(clientId);
+            tokenDTO.setTokenAuth(getbase64EncodedTokenAouth(tokenDTO.getTokenAuth()));
+            tokenList.add(tokenDTO);
+            newSpDto.setSpTokenList(tokenList);
+            callTokenPool(newSpDto);
+        }
+    }
+
+    @Override
+    public boolean handleResponse(MessageContext messageContext) {
+        return true;
+    }
+
+    private void callTokenPool(final AddNewSpDTO newSpDto) {
+        ExecutorService executorService = Executors.newFixedThreadPool(1);
```

Comment:
use executor service effectively. This gets executed per request.

Reply:
This Method is only trigger if Token pool is enabled as per requirement
 if (propertyMap.get(TOKEN_POOL_ENABLED).equals("true")) {
            if (log.isDebugEnabled()) {
                log.debug("TOken Pool Service Enabled for Insertion");
            }
            AddNewSpDTO newSpDto = new AddNewSpDTO();
            ArrayList<TokenDTO> tokenList = new ArrayList<TokenDTO>();
            newSpDto.setOwnerId(clientId);
            tokenDTO.setTokenAuth(getbase64EncodedTokenAouth(tokenDTO.getTokenAuth()));
            tokenList.add(tokenDTO);
            newSpDto.setSpTokenList(tokenList);
            callTokenPool(newSpDto);
        }